12.0 add customer device manager finalize

fcayre opened the pull request.

fcayre committed:
Add an action to primary service storable config to copy the image

The button simply calls a method that copies the service image
into the storable template image.

The template image is copied to the variant's by the corresponding
core module during this operation.

A user error is raised if the button is pressed while the service
product still has no image.

fcayre committed:
customer_team_manager and customer_device_manager i18n update and complete FR/DE t10n

fcayre commented:
@loganSIO les tests plantent ici (comme partout en v12 désormais) à cause d'un souci dans les tests de commown_self_troubleshooting lié à chromium (tests javascript tours), donc pas lié à ton dév (ni à mon ajout).

fcayre committed:
Refactor test to help code reuse in an upcoming new test

Also remove unused variable.

fcayre committed:
Simplify and improve assignment test a bit

The get_assignments method no longer filters its results by company,
we instead check what company the result is related to. This simplifies
the implementation while improving the test expectation checks.

fcayre committed:
Do not reveal contract name to customers when it is not theirs

When the contract associated with a device is not related to a partner's
company, it is no longer displayed.
To implement this bug fix, the contract_name field had to be turned into
a (still non-stored) compute instead of a simple related, so that we can
test the device contract's partner to decide if the contract name should
be exposed or not.

fcayre committed:
Improve assignment history displayed name

It now contains the partner's displayed name and assignment date.

We do no longer protect against a missing device_id as this field is
required, which simplifies the implementation and test.

fcayre committed:
Fix incorrect date of the first assignment history item

If an assignment is create with an assignment_date in the past, the
first history item must be created with this date instead of `now`.

This is now enforced using a context variable set in the assignment
`create` method.

fcayre committed:
Display the partner name on the assignment history tree view

... even if we group by partner by default, this is a precious information
(in particular if the user removes the default grouping).

fcayre committed:
Replace device assignment 'product_id' by 'product_name' to make it customer searchable

The 'product.product' model is indeed not readable by customers so
its name is not searchable. We do not need more than its name so we
store the product name directly, which makes it searchable.

Note the product name is computed when the assignment is created and
does never change, so we should never have a permission read error
when it is computed, hence the absence of sudo_compute=True.

fcayre committed:
Store device assignment's contract_name field to make it searchable

fcayre committed:
Add a link to the device manager in the portal home

fcayre committed:
Review customer manager menu names and permissions

Root menu is now called "Management space". It is accessible to anyone
who as access to one of its components: members of group_customer_admin
(who manage users and teams) and of group_customer_device_assigner
(who manage devices).

fcayre committed:
customer management space portal UI fixes

- A typo is fixed (managment becomes management)
- The portal section appears when at least one link inside is accessible
  by current user: either users or devices managers.
- Links to user/team manager and device manager are displayed to the
  right group members

fcayre committed:
Fix typo in fleet management role name

fcayre committed:
Add team read permission to portal users of the same company

fcayre committed:
Selectively display portal access management buttons on customer admin manager partner view

Only members of the sales_team.group_sale_manager and
customer_team_manager.group_customer_admin groups are supposed to use
them.

fcayre committed:
Security fix: protect partner revoke action

Before the patch, it was possible for any portal user to use the action to
revoke portal access to a collegue (same company user).

It is now impossible and tested.

fcayre committed:
Do not hide partner's team field: teams are now also readable by portal users

fcayre committed:
Disallow customer admin write permission to companies

... and fix latent security bug: write permission to the object was not checked
correctly, as only the read permission and the user groups were checked.

Now we disallow the write permission on companies, so we need to fix this bug as
the read and write rules become different (read permission to companies is OK).

Added a test to make things clearer.

fcayre committed:
Suppress almost all content of the customer partner view for companies

It's not easy to disable this view for companies, that are accessible from device
assignments at least (and are common as assignment partners), so we hide almost
everything to avoid access errors if customers try to use this view to modify a
company.

fcayre committed:
Filter out companies from portal access customer wizard

fcayre committed:
Fix portal links to customer manager pages

fcayre committed:
Hide useless fields in the filter UI widget

This makes usage by customer much simpler.

fcayre committed:
Remove the Favorites menu from the search view when user is a customer

This is a bit radical, but this feature is not that useful in the
context of customer dedicaded UIs and keeping it would require to hide
some sub-functionnalities (namely: share a favorite with users and save
a favorite to the dashboard).

fcayre committed:
Detect when current portal user is a customer instead of an admin customer

To adapt the interface client-side, this is enough to know current user
is a customer (= not in internal user). This will help adapt the search
view in a future commit as it helped customizing the export data view
in passed ones.

fcayre committed:
Group customer user menu items under a single menu

This is more consistent with the devices menu items.

Honeyxilia commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -57,16 +57,10 @@ def _ref(name):
 
         self.internal_user = self.env.ref("base.group_user")
 
-    def get_assignments(self, lot, company):
```
Erreur de message de commit : 
- "method does no more filter" -> "method no longer filters"

Honeyxilia commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -84,7 +84,7 @@ class DeviceAssignment(models.Model):
 
     contract_name = fields.Char(
         string="Contract",
-        related="device_id.contract_id.name",
```
Erreur dans le commit : 
- "it is no more displayed" -> "it's no longer displayed"
- "the contract name **could** be exposed" -> "the contract name **should** be exposed" (plus idiomatique que incorrect)

Honeyxilia commented on customer_device_manager/tests/test_stock_move.py:
```diff
@@ -69,13 +69,14 @@ def _send(self, lot, orig, dest):
+    def test_device_assignment_lifecycle_same_customer(self):
         """
-        Test the lifecycle of a device assignment:
+        Test a real lifecycle of a device assignment:
```
Je pense qu'il serait plus pertinent de dire "Test a typical lifecycle" plutôt que "Test a real lifecycle"

Honeyxilia commented on customer_device_manager/models/device_assignment_history.py:
```diff
@@ -34,6 +34,12 @@ class DeviceAssignmentHistory(models.Model):
     )
 
     def name_get(self):
-        return [
```
Message de commit : "We *do no more* protect" -> "We **no longer** protect"

Honeyxilia commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -89,12 +89,28 @@ class DeviceAssignment(models.Model):
+            _self = self.with_context(forced_assignment_history_date=_date)
+
+        # Restore the original context in the returned result to avoid any side effect:
+        return super(DeviceAssignment, _self).create(values).with_context(orig_context)
```
Ici, est-ce que `_self` n'hérite pas déjà du contexte du `self` de base ?

fcayre replied:
Oui il en hérite mais on enlève ainsi `forced_assignment_history_date` (et il faut sinon il y a des effets de bord car elle reste dans le contexte de la valeur retournée par create).

Honeyxilia commented on customer_device_manager/tests/test_device_assignment.py:
```diff
@@ -87,19 +87,24 @@ def test_date_update_history_date(self):
+            "test-lot - Joel Willis (2023-01-01 10:00:00)",
+        )
+
+        assignment = self.env[assignment._name].browse(assignment.id)
```
J'ai l'impression que cette instruction n'est pas très utile - quel est son objectif ?
Si c'était pour mettre à jour le champ display_name, est-ce qu'il n'aurait pas été plus clair d'invalider le cache ?

fcayre replied:
Bien vu, je l'enlève, je pense que c'est un résidu d'une ancienne implémentation, merci !

Honeyxilia commented on customer_device_manager/models/device_assignment.py:
```diff
@@ -74,11 +74,10 @@ class DeviceAssignment(models.Model):
         store=False,
```
petite typo de commit : "when *is* is computed" -> "when *is* is computed

Honeyxilia reviewed (commented):
Quelques questionnement sur certaines opérations, mais sinon c'est OK pour moi c:

fcayre commented:
@Honeyxilia remarques prises en compte, merci !

fcayre committed:
[customer_team_manager] Add an Help URL action targetting a forum article

fcayre committed:
[customer_device_manager] Add an Help URL action targetting a forum article

fcayre committed:
The partner of a device assignment located at commown is now readonly in the UI

... otherwise customers trying to update it will get an error.

Internal users do no need to modify it either, so no group condition is added to
the readonly domain.

fcayre commented:
/ocabot merge patch
